src/main/java/org/opensearch/knn/index/query/lucene/LuceneEngineKnnVectorQuery.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.query.lucene;
+
+import lombok.AllArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Weight;
+
+import java.io.IOException;
+
+/**
+ * LuceneEngineKnnVectorQuery is a wrapper around a vector queries for the Lucene engine.
+ * This enables us to defer rewrites until weight creation to optimize repeated execution
+ * of Lucene based k-NN queries.
+ */
+@AllArgsConstructor
+@Log4j2
+public class LuceneEngineKnnVectorQuery extends Query {
+    private final Query luceneQuery;
+
+    /*
+      Prevents repeated rewrites of the query for the Lucene engine.
+    */
+    @Override
+    public Query rewrite(IndexSearcher indexSearcher) {
+        return this;
+    }
+
+    /*
+       Rewrites the query just before weight creation.
+     */
+    @Override
+    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
+        Query rewrittenQuery = luceneQuery.rewrite(searcher);
+        return rewrittenQuery.createWeight(searcher, scoreMode, boost);
+    }
+
+    @Override
+    public String toString(String s) {
```

Comment:
some of methods from java Object can be generated by Lombok annotations, e.g. @ToString and @@EqualsAndHashCode

Reply:
For query class we should just use the delegation approach here rather than using lombok. That would be my recommendation.

Reply:
ah, makes sense, delegating here makes more sense